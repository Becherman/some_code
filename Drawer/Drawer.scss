$shadow: $box-shadow-2;

:local(.drawer) {
  padding: 0;
}

:local(.layout) {
  display: flex;
  justify-content: space-between;
}

:local(.backdrop) {
  height: 100vh;
  width: 100%;
  top: 0;
  left: 0;
  position: fixed;
  opacity: 0;
  pointer-events: none;
  transition: opacity 0.2s ease-in;
  z-index: 9;
  background-color: #000;
}

:local(.backdropIsOpen) {
  opacity: 0.2;
  pointer-events: all;
}

:local(.mobileSidebar) {
  position: fixed;
  right: 0;
  top: 0;
  width: 70%;
  height: 100vh;
  background: $white;
  z-index: 10;
  opacity: 0;
  transition: all 0.4s ease-out;
  transform: translateX(100%);
}

:local(.sidebarIsOpen) {
  transform: translateX(0);
  opacity: 1;
}

:local(.sidebar) {
  display: none;
}

:local(.main) {
  width: 100%;
}

@include for-desktop {
  :local(.drawer) {
    padding: 16px;
  }

  :local(.sidebar) {
    display: block;
    width: 25%;
    background: #fff;
    box-shadow: $shadow;
  }

  :local(.main) {
    display: block;
    width: 73%;
  }
  
  :local(.mobileSidebar),
  :local(.toggler),
  :local(.backdrop) {
    display: none;
  }
}
