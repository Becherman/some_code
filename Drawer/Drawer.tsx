import * as React from 'react';
import classNames from 'classnames';

import styles from './Drawer.scss';

interface Props {
  sidebar(sidebarHandler: () => void): React.ReactNode;
  main(sidebarHandler: () => void): React.ReactNode;
}

interface State {
  sidebarIsOpen: boolean;
}
export class Drawer extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      sidebarIsOpen: false,
    };
  }

  toggleSidebar = (): void => {
    this.setState(state => ({
      sidebarIsOpen: !state.sidebarIsOpen,
    }));
  };

  render() {
    const { sidebar, main } = this.props;
    const { sidebarIsOpen } = this.state;

    return (
      <div className={styles.drawer}>
        {/** Backdrop */}
        <div
          onClick={this.toggleSidebar}
          className={classNames({
            [styles.backdrop]: true,
            [styles.backdropIsOpen]: sidebarIsOpen,
          })}
        />

        {/** Mobile Sidebar */}
        <div
          className={classNames({
            [styles.mobileSidebar]: true,
            [styles.sidebarIsOpen]: sidebarIsOpen,
          })}
        >
          {sidebar(this.toggleSidebar)}
        </div>

        <div className={styles.layout}>
          {/** Desktop Sidebar */}
          <div className={styles.sidebar}>{sidebar(this.toggleSidebar)}</div>

          {/** Main Content */}
          <div className={styles.main}>{main(this.toggleSidebar)}</div>
        </div>
      </div>
    );
  }
}
