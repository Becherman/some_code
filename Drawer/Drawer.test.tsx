import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Drawer } from './Drawer';

const Component = () => <div></div>;

describe('[Components] => Drawer', () => {
  it('should render without errors', () => {
    const wrapper = shallow(
      <Drawer sidebar={() => <Component />} main={() => <Component />} />,
    );

    expect(wrapper.find('.drawer').children()).toHaveLength(3);
  });

  it('should render a sidebar, main content section and a backdrop', () => {
    const wrapper = shallow(
      <Drawer sidebar={() => <Component />} main={() => <Component />} />,
    );

    expect(wrapper.find('.sidebar').exists()).toBe(true);
    expect(wrapper.find('.main').exists()).toBe(true);
    expect(wrapper.find('.backdrop').exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should update a classnames after clicking on a backdrop', () => {
    const wrapper = shallow(
      <Drawer sidebar={() => <Component />} main={() => <Component />} />,
    );
    const backdrop = wrapper.find('.backdrop');

    expect(wrapper.find('.mobileSidebar').hasClass('sidebarIsOpen')).toBe(
      false,
    );
    expect(toJson(wrapper)).toMatchSnapshot();

    backdrop.simulate('click');

    expect(wrapper.find('.mobileSidebar').hasClass('sidebarIsOpen')).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should update class names when the state will change', () => {
    const wrapper = shallow(
      <Drawer sidebar={() => <Component />} main={() => <Component />} />,
    );

    expect(wrapper.find('.mobileSidebar').hasClass('sidebarIsOpen')).toBe(
      false,
    );
    expect(wrapper.find('.backdrop').hasClass('backdropIsOpen')).toBe(false);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.setState({ sidebarIsOpen: true });

    expect(wrapper.find('.mobileSidebar').hasClass('sidebarIsOpen')).toBe(true);
    expect(wrapper.find('.backdrop').hasClass('backdropIsOpen')).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
