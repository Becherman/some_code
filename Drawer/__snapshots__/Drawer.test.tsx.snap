// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Components] => Drawer should render a sidebar, main content section and a backdrop 1`] = `
<div
  className="drawer"
>
  <div
    className="backdrop"
    onClick={[Function]}
  />
  <div
    className="mobileSidebar"
  >
    <Component />
  </div>
  <div
    className="layout"
  >
    <div
      className="sidebar"
    >
      <Component />
    </div>
    <div
      className="main"
    >
      <Component />
    </div>
  </div>
</div>
`;

exports[`[Components] => Drawer should update a classnames after clicking on a backdrop 1`] = `
<div
  className="drawer"
>
  <div
    className="backdrop"
    onClick={[Function]}
  />
  <div
    className="mobileSidebar"
  >
    <Component />
  </div>
  <div
    className="layout"
  >
    <div
      className="sidebar"
    >
      <Component />
    </div>
    <div
      className="main"
    >
      <Component />
    </div>
  </div>
</div>
`;

exports[`[Components] => Drawer should update a classnames after clicking on a backdrop 2`] = `
<div
  className="drawer"
>
  <div
    className="backdrop backdropIsOpen"
    onClick={[Function]}
  />
  <div
    className="mobileSidebar sidebarIsOpen"
  >
    <Component />
  </div>
  <div
    className="layout"
  >
    <div
      className="sidebar"
    >
      <Component />
    </div>
    <div
      className="main"
    >
      <Component />
    </div>
  </div>
</div>
`;

exports[`[Components] => Drawer should update class names when the state will change 1`] = `
<div
  className="drawer"
>
  <div
    className="backdrop"
    onClick={[Function]}
  />
  <div
    className="mobileSidebar"
  >
    <Component />
  </div>
  <div
    className="layout"
  >
    <div
      className="sidebar"
    >
      <Component />
    </div>
    <div
      className="main"
    >
      <Component />
    </div>
  </div>
</div>
`;

exports[`[Components] => Drawer should update class names when the state will change 2`] = `
<div
  className="drawer"
>
  <div
    className="backdrop backdropIsOpen"
    onClick={[Function]}
  />
  <div
    className="mobileSidebar sidebarIsOpen"
  >
    <Component />
  </div>
  <div
    className="layout"
  >
    <div
      className="sidebar"
    >
      <Component />
    </div>
    <div
      className="main"
    >
      <Component />
    </div>
  </div>
</div>
`;
